src/promptflow/promptflow/executor/_tool_resolver.py

```diff
@@ -107,7 +111,78 @@ def _convert_to_assistant_definition(self, assistant_definition_path: str, input
+                resolved_tools[function_tool.name] = function_tool
+            else:
+                raise UnsupportedAssistantToolType(
+                    message_format="Unsupported assistant tool type: {tool_type}",
+                    tool_type=tool["type"],
+                    target=ErrorTarget.EXECUTOR,
+                )
+        assistant_definition.tool_invoker = AssistantToolInvoker(resolved_tools)
+
+    def _load_tool_as_function(self, tool: dict):
+        node, predefined_inputs = self._create_node_for_assistant_tool(tool)
+        resolved_tool = self.resolve_tool_by_node(node, convert_input_types=True)
+        func_name = resolved_tool.definition.function
+        definition = self._generate_tool_definition(func_name, resolved_tool.definition.description, predefined_inputs)
+        if resolved_tool.node.inputs:
+            inputs = {name: value.value for name, value in resolved_tool.node.inputs.items()}
+            func = partial(resolved_tool.callable, **inputs)
+        else:
+            func = resolved_tool.callable
+        return AssistantTool(name=func_name, openai_definition=definition, func=func)
+
+    def _create_node_for_assistant_tool(self, tool: dict):
+        predefined_inputs = {}
+        for input_name, value in tool.get("predefined_inputs", {}).items():
+            predefined_inputs[input_name] = InputAssignment.deserialize(value)
+        node = Node(
+            name="assistant_node",
+            tool="assistant_tool",
+            inputs=predefined_inputs,
+            source=ToolSource.deserialize(tool["source"]) if "source" in tool else None,
+            type=ToolType.PYTHON if "tool_type" in tool and tool["tool_type"] == "python" else None,
+        )
+        return node, list(predefined_inputs.keys())
+
+    def _generate_tool_definition(self, func_name: str, description: str, predefined_inputs: list) -> dict:
+        try:
+            to_openai_type = {
+                "str": "string",
+                "int": "number",
+                "float": "number",
+                "bool": "boolean",
+                "list": "array",
+                "dict": "object",
+            }
+            description, params = DocstringParser.parse(description)
+            for input in predefined_inputs:
+                if input in params:
+                    params.pop(input)
+            for _, param in params.items():
+                param["type"] = to_openai_type[param["type"]] if param["type"] in to_openai_type else param["type"]
+
+            return {
+                "type": "function",
+                "function": {
+                    "name": func_name,
+                    "description": description,
+                    "parameters": {"type": "object", "properties": params, "required": list(params.keys())},
+                },
+            }
+        except Exception as e:
+            raise FailedToParseAssistantTool(func_name=func_name) from e
```

Comment:
If we raise the exception here, all the tools without doc string cannot be used?

Reply:
right, currently it is.  I think we used docstring to parse parameter +description. I created a working item to improve this:

Reply:
I think this exception would be deprecate or downgrade to be warning once we get independent of docstring for function parse